core/src/chain_id.rs

```diff
@@ -0,0 +1,160 @@
+use mina_p2p_messages::v2::{
+    MinaBaseProtocolConstantsCheckedValueStableV1, StateHash, UnsignedExtendedUInt32StableV1,
+};
+use multihash::{Blake2b256, Hasher};
+use time::macros::format_description;
+use time::OffsetDateTime;
+
+use std::fmt::{self, Debug, Display, Formatter};
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct ChainId([u8; 32]);
+
+fn u8_slice_as_hex(slice: &[u8]) -> String {
+    slice.iter().map(|b| format!("{:02x}", b)).collect()
+}
+
+fn md5_hash(data: u8) -> String {
+    let mut hasher = md5::Context::new();
+    hasher.consume(data.to_string().as_bytes());
+    let hash: Md5 = *hasher.compute();
+    u8_slice_as_hex(&hash)
+}
+
+type Md5 = [u8; 16];
+
+fn hash_genesis_constants(
+    constants: &MinaBaseProtocolConstantsCheckedValueStableV1,
+    tx_pool_max_size: &UnsignedExtendedUInt32StableV1,
+) -> [u8; 32] {
+    let mut hasher = Blake2b256::default();
+    let genesis_timestamp = OffsetDateTime::from_unix_timestamp_nanos(
+        (constants.genesis_state_timestamp.0 .0 .0 * 1000000) as i128,
+    )
+    .unwrap();
+    let time_format =
+        format_description!("[year]-[month]-[day] [hour]:[minute]:[second].[subsecond digits:6]Z");
+    let input = format!(
+        "{}{}{}{}{}{}",
+        constants.k.to_string(),
+        constants.slots_per_epoch.to_string(),
+        constants.slots_per_sub_window.to_string(),
+        constants.delta.to_string(),
+        tx_pool_max_size.to_string(),
+        genesis_timestamp.format(&time_format).unwrap(),
+    );
+    hasher.update(input.as_bytes());
+    hasher.finalize().try_into().unwrap()
+}
+
+impl ChainId {
+    pub fn compute(
+        constraint_system_digests: &[Md5],
+        genesis_state_hash: &StateHash,
+        genesis_constants: &MinaBaseProtocolConstantsCheckedValueStableV1,
+        protocol_transaction_version: u8,
+        protocol_network_version: u8,
+        tx_max_pool_size: &UnsignedExtendedUInt32StableV1,
+    ) -> ChainId {
+        let mut hasher = Blake2b256::default();
+        let constraint_system_hash = constraint_system_digests
+            .iter()
+            .map(|md5| u8_slice_as_hex(md5))
+            .collect::<Vec<String>>()
+            .join("");
```

Comment:
Probably something like `Iterator::reduce` will be more appropriate then `collect/join`, without allocating a vector.